api/grpcserver/v2alpha1/malfeasance.go

```diff
@@ -65,36 +68,56 @@ func (s *MalfeasanceService) List(
 		return nil, status.Error(codes.InvalidArgument, "limit must be set to <= 100")
 	}
 
-	ops, err := toMalfeasanceOps(request)
+	legacyCount, err := identities.CountMalicious(s.db)
 	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	proofs := make([]*spacemeshv2alpha1.MalfeasanceProof, 0, request.Limit)
-	if err := identities.IterateOps(s.db, ops, func(id types.NodeID, proof []byte, received time.Time) bool {
-		rst := toProof(ctx, s.info, id, proof)
-		if rst == nil {
-			return true
+	switch {
+	case request.Offset+request.Limit < legacyCount: // only legacy proofs
+		proofs, err := fetchLegacyFromDB(ctx, s.db, s.infoLegacy, request)
+		if err != nil {
+			return nil, err
 		}
-		proofs = append(proofs, rst)
-		return true
-	}); err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+		return &spacemeshv2alpha1.MalfeasanceList{Proofs: proofs}, nil
+	case request.Offset >= legacyCount: // only new proofs
+		request.Offset -= legacyCount
+		proofs, err := fetchFromDB(ctx, s.db, s.info, request)
+		if err != nil {
+			return nil, err
+		}
+		return &spacemeshv2alpha1.MalfeasanceList{Proofs: proofs}, nil
+	default: // both legacy and new proofs
+		legacyProofs, err := fetchLegacyFromDB(ctx, s.db, s.infoLegacy, request)
```

Comment:
This may cause some issues while we're still accepting v1 malfeasance proofs.
Imagine that a new v1 malfeasance proof is processed and stored in the database after we call `identities.CountMalicious(...)` call but before `fetchLegacyFromDB(...) call. In this case, more than `request.Limit` entries may be retrieved (v1+v2).
This could be fixed by wrapping database accesses in a read transaction (non-immediate)

Reply:
I don't think it would matter since the limit of the query for `fetchFromDB` is adjusted based on what was returned from `fetchLegacyFromDB`. So even if between counting and reading the data new proofs were inserted, the number of returned proofs should still not exceed the limit. However I still wrapped the queries into a single TX for consistency.